src/solver/mod.rs

```diff
@@ -35,17 +35,57 @@ struct AddClauseOutput {
 }
 
 /// Describes the problem that is to be solved by the solver.
-#[derive(Default)]
-pub struct Problem {
-    /// The requirements that _must_ have one candidate solvable be included in the
+///
+/// This struct follows the builder pattern and can have its fields set by one of the available
+/// setter methods.
+pub struct Problem<S> {
```

Comment:
Im wondering what makes soft_requirements special conpared to requirements. Why can we use a Vec for the soft requirements but not for the other fields?

Reply:
The `soft_requirements` are only iterated over once in the `Solver::solve` method, unlike `requirements` and `constraints` which need to be stored in the `root_requirements` and `root_constraints` fields of the `Solver` struct (so they can be cloned and specified in the `Dependencies` of the root solvable whenever it is requested).

Reply:
We could very well choose to use an iterator for `requirements` and `constraints` as well, but they would end up being `collect`ed into a `Vec` to be stored in the `root_requirements` and `root_constraints` fields of the `Solver` struct anyway.